index.d.ts

```diff
@@ -16,6 +20,46 @@ export interface Options {
 	readonly stopOnError?: boolean;
 }
 
+export interface OngoingMappingsStopOptions {
+	/**
+	Whether or not to remove all holes from the result array (caused by pending mappings).
+
+	@default false
+	*/
+	readonly collapse?: boolean;
+
+	/**
+	Value to use as immediate result for pending mappings, replacing holes from the result array.
+
+	This option is ignored if `collapse` is set to `true`.
+
+	@default undefined
+	*/
+	readonly fillWith?: unknown;
+}
+
+export interface StopOptions<NewElement> {
+	/**
+	Value to provide as result for this iteration.
+
+	@default undefined
+	*/
+	readonly value?: NewElement;
+
+	/**
+	Options to configure what `pMap` must do with any concurrent ongoing mappings at the moment `stop` is called.
+	*/
+	readonly ongoingMappings?: OngoingMappingsStopOptions;
```

Comment:
I wonder if there's any way we could flatten the options object:

- value
- collpasePending
- fillPending

Needs better names.

Reply:
I think it makes sense to group them. In the future, if support for canceling promises and `AbortController` is added, they might probably need options that could also be added to this group.

Do you also want better names for the way I've currently done it? Or you said "better names" only for your suggestion itself?

Reply:
Maybe something like this:

- `value` Value to resolve with
- `defaultValue` Value to set when the promise hasn't been run
	What if the consumer wants it to be an empty value? Maybe this could be in the form of passing a `pMap.stop.empty` symbol
- `stopRunning` Whether currently running promises should be allowed to finish
- `removeIncomplete` Whether empty items in the resulting array where there should have been values if all the promises had run should be removed.

Reply:
I still think the non-flattened way is better, because `value` is about the current iteration, while everything else is about *configuring what happens to the rest of the mapping itself*. Maybe instead of `ongoingMappings`, one of:

* `rest`
* `restConfig`
* `restOptions`
* `whatToDoWithOngoingMappings`
* `mappingsStillRunning`
* `inProgressBehavior`
* `inProgressConfig`

Reply:
> What if the consumer wants it to be an empty value? Maybe this could be in the form of passing a `pMap.stop.empty` symbol

I think this particular behavior could be left for another PR. I've never seen good uses for holes in arrays, they're an obscure feature of the language, if I'm not mistaken TypeScript for example pretends they don't exist. Sindre also said it [here](https://github.com/sindresorhus/p-map/issues/38#issuecomment-868983977). Although if you found a good use case for them, I'd be interested in knowing more.

Reply:
Alright, we can stick with the nested option. I'm not really sold on most of these names though. The problem is that it applies to any mappings, not just running ones, but also pending ones. For example, if the `concurrency` option is used, not all remaining mappings are executing when it's stopped. I think the only one that is usable is `rest`, so I guess we can go with that.